crates/proof-of-sql/src/base/database/column_operation.rs

```diff
@@ -0,0 +1,371 @@
+        let precision_value = left_precision_value + right_precision_value + 1;
+        let precision = Precision::new(precision_value).map_err(|_| {
+            ColumnOperationError::DecimalConversionError(DecimalError::InvalidPrecision(format!(
+                "Required precision {} is beyond what we can support",
+                precision_value
+            )))
+        })?;
+        let left_scale = lhs.scale().expect("Numeric types have scale");
+        let right_scale = rhs.scale().expect("Numeric types have scale");
+        let scale = left_scale.checked_add(right_scale).ok_or(
+            ColumnOperationError::DecimalConversionError(DecimalError::InvalidScale(
+                left_scale as i16 + right_scale as i16,
+            )),
+        )?;
+        Ok(ColumnType::Decimal75(precision, scale))
+    }
+}
+
+/// Add two slices of the same length.
+///
+/// We do not check for length equality here
+fn add_slices<T>(lhs: &[T], rhs: &[T]) -> Vec<T>
+where
+    T: Add<Output = T> + Copy,
+{
+    lhs.iter().zip(rhs).map(|(l, r)| *l + *r).collect()
+}
+
+/// Add two slices of the same length, casting the left slice to the type of the right slice.
+///
+/// We do not check for length equality here
+fn add_slices_with_casting<SmallerType, LargerType>(
+    numbers_of_smaller_type: &[SmallerType],
+    numbers_of_larger_type: &[LargerType],
+) -> Vec<LargerType>
+where
+    SmallerType: Copy,
+    LargerType: From<SmallerType> + Add<Output = LargerType> + Copy,
+{
+    numbers_of_smaller_type
+        .iter()
+        .zip(numbers_of_larger_type)
+        .map(|(l, r)| LargerType::from(*l) + *r)
+        .collect()
+}
+
+/// Add two numerical slices as decimals.
+fn add_decimal_columns<S, T0, T1>(
+    lhs: &[T0],
+    rhs: &[T1],
+    left_column_type: ColumnType,
+    right_column_type: ColumnType,
+) -> ColumnOperationResult<(Precision, i8, Vec<S>)>
+where
+    S: Scalar + core::convert::From<T0> + core::convert::From<T1>,
+    T0: Copy,
+    T1: Copy,
+{
+    let new_column_type =
+        try_add_subtract_column_types(left_column_type, right_column_type, BinaryOperator::Add)?;
+    let new_precision_value = new_column_type
+        .precision_value()
+        .expect("numeric columns have precision");
+    let new_scale = new_column_type.scale().expect("numeric columns have scale");
+    let left_upscale = new_scale
+        - left_column_type
+            .scale()
+            .expect("numeric columns have scale");
+    let right_upscale = new_scale
+        - right_column_type
+            .scale()
+            .expect("numeric columns have scale");
+    // One of left_scale and right_scale is 0 so we can avoid scaling when unnecessary
+    let scalars: Vec<S> = if right_upscale == 0 {
+        lhs.iter()
+            .zip(rhs)
+            .map(|(l, r)| -> ColumnOperationResult<S> {
+                Ok(scale_scalar(S::from(*l), left_upscale)? + S::from(*r))
```

Comment:
`scale_scalar` has to compute the 10^left_upscale each time here, which will by far dominate the conversion.
If you do the `upscale_factor = 10^left_upscale` outside of the iter, you can simply do:
```suggestion
S::from(*l) * upscale_factor + S::from(*r)
```
and even avoid the `Ok`.